### PR TITLE
Test if reserved filenames are allowed in Windows

### PR DESCRIPTION
**Pull request by eholk**
Recent versions of Windows have removed the limitation on filenames like `aux` or `con`. This change allows the `package::reserved_windows_name` to still pass by first trying to create a file with a reserved name to see if Windows supports it. If so, it skips the rest of the test. Otherwise, we keep the same behavior as before.

### CODE REVIEW

File: src/cargo/util/restricted_names.rs [hunk 1]
```diff
@@ -97,3 +97,59 @@ pub fn is_windows_reserved_path(path: &Path) -> bool {
 pub fn is_glob_pattern<T: AsRef<str>>(name: T) -> bool {
     name.as_ref().contains(&['*', '?', '[', ']'][..])
 }
+
+/// Returns true if names such as aux.* are allowed.
+///
+/// Traditionally, Windows did not allow a set of file names (see `is_windows_reserved_name`
```
Comment (weihanglo) [inline src/cargo/util/restricted_names.rs:103]: Should be `is_windows_reserved`?

**Review by weihanglo — COMMENTED** [review]
Hi. If I want to verify the behavior of this patch. What environment should I prepare? Specifically, can you provide some official doc about this change? Thanks!

File: src/cargo/util/restricted_names.rs (hunk 1 above)
Comment (eholk) [reply src/cargo/util/restricted_names.rs]: Yes, fixed.

**Comment by eholk** [comment]
> Hi. If I want to verify the behavior of this patch. What environment should I prepare? Specifically, can you provide some official doc about this change? Thanks!

I'm on the Windows 11 Insiders Beta channel, and winver says it's version 21H2 (OS Build 22000.466).

Unfortunately, I haven't been able to find official documentation about the change, but I did confirm the behavior with the Windows dev who made the change.

**Comment by joshtriplett** [comment]
Could you post a transcript of the test failure (including with nocapture so we can see the actual test output)?

**Comment by eholk** [comment]
Sure, here's the output from `cargo test -- package::reserved_windows_name --nocapture` on `master`:

```
    Finished test [unoptimized + debuginfo] target(s) in 0.18s
     Running unittests src/cargo/lib.rs (target\debug\deps\cargo-9040197b200da57a.exe)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 34 filtered out; finished in 0.00s

     Running tests\build-std\main.rs (target\debug\deps\build_std-9cc2dc147da3e21f.exe)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 3 filtered out; finished in 0.00s

     Running tests\internal.rs (target\debug\deps\internal-870b738d2af5cbc1.exe)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 2 filtered out; finished in 0.00s

     Running tests\testsuite\main.rs (target\debug\deps\testsuite-508cfd0591301aff.exe)

running 1 test
running `C:\Users\ericholk\repo\cargo\target\debug\cargo.exe package`
thread 'package::reserved_windows_name' panicked at '
test failed running `C:\Users\ericholk\repo\cargo\target\debug\cargo.exe package`
error: process exited with code 0 (expected 101)
--- stdout

--- stderr
warning: manifest has no documentation, homepage or repository.
See https://doc.rust-lang.org/cargo/reference/manifest.html#package-metadata for more info.
   Packaging foo v0.0.1 (C:\Users\ericholk\repo\cargo\target\tmp\cit\t0\foo)
    Updating `dummy-registry` index
   Verifying foo v0.0.1 (C:\Users\ericholk\repo\cargo\target\tmp\cit\t0\foo)
 Downloading crates ...
  Downloaded bar v1.0.0 (registry `dummy-registry`)
   Compiling bar v1.0.0
   Compiling foo v0.0.1 (C:\Users\ericholk\repo\cargo\target\tmp\cit\t0\foo\target\package\foo-0.0.1)
    Finished dev [unoptimized + debuginfo] target(s) in 0.40s
', tests\testsuite\package.rs:1939:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test package::reserved_windows_name ... FAILED

failures:

failures:
    package::reserved_windows_name

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 2345 filtered out; finished in 0.67s

error: test failed, to rerun pass '--test testsuite'
```

**Comment by joshtriplett** [comment]
So, reading that test case, I don't think that's a test that should pass in general, even if the platform doesn't actually reserve such filenames. The logic that checks for `aux.rs` and avoids unpacking the file should not depend on what Windows does, at least not until we stop supporting all versions of Windows with that limitation.

I think, rather than changing the test so its behavior depends on the OS, this should instead change the implementation of unpacking to ensure that it catches reserved filenames on all versions of Windows.

**Comment by eholk** [comment]
That seems reasonable. If I remember right, the code for publishing packages tries to prevent you from publishing packages that couldn't be unpacked on all OSes supported by Cargo.

In that case though, I think we should also remove the `#[cfg(windows)]` on this test.

**Comment by joshtriplett** [comment]
> If I remember right, the code for publishing packages tries to prevent you from publishing packages that couldn't be unpacked on all OSes supported by Cargo, so it makes sense to have similar behavior for unpacking packages.

Exactly.



> In that case though, I think we should also remove the `#[cfg(windows)]` on this test.

Agreed.

**Comment by eholk** [comment]
I decided to close this PR in favor of #10461 which cleanly does what was decided in the discussion here.

**Comment by eholk** [comment]
Reopening this since there's disagreement about whether we want this change or the one in #10461. We only need one of these to land though.

**Comment by joshtriplett** [comment]
Per discussion on Zulip, this approach seems fine to me, as long as it doesn't actually use the same code in both this check and the code being tested (which would be circular).

In that spirit, and in the spirit of not providing this as a public API of Cargo, could you move this function into the test code rather than the util module?

**Comment by eholk** [comment]
> In that spirit, and in the spirit of not providing this as a public API of Cargo, could you move this function into the test code rather than the util module?

Yup, that's done now! I moved it to the `cargo_test_support` crate, which required adding `winapi` as a dependency there too. Hopefully that's okay!

**Comment by joshtriplett** [comment]
Seems reasonable to me. Does anyone see an issue with adding that dependency to the test support crate?

File: crates/cargo-test-support/src/paths.rs [hunk 2]
```diff
@@ -287,3 +287,54 @@ pub fn sysroot() -> String {
...
     sysroot.trim().to_string()
 }
+
+/// Returns true if names such as aux.* are allowed.
+///
+/// Traditionally, Windows did not allow a set of file names (see `is_windows_reserved`
+/// for a list). More recent versions of Windows have relaxed this restriction. This test
+/// determines whether we are running in a mode that allows Windows reserved names.
+#[cfg(windows)]
+pub fn windows_reserved_names_are_allowed() -> bool {
```
Comment (ehuss) [inline crates/cargo-test-support/src/paths.rs:297]: Can this add something like:

```rust
if is_ci() { return false; }
```

I'd like to ensure the tests are still run on CI until such time that we need to migrate.

**Review by ehuss — COMMENTED** [review]
I'm sorry for the super long delay. I didn't have a Windows 11 system handy. I would be happy to get this merged with a small change.

File: crates/cargo-test-support/src/paths.rs (hunk 2 above)
Comment (eholk) [reply crates/cargo-test-support/src/paths.rs:297]: Sure, I just pushed the change that does this.

**Comment by ehuss** [comment]
Thanks!

@bors r+

**Pull request merged**